### PR TITLE
Update related_descriptors.py

### PR DESCRIPTION
**Pull request by adontz**
Avoid unnecessary database UPDATEs if objects are already related.

### CODE REVIEW

**Comment by charettes** [comment]
Hello @adontz, [per our contribution guidelines](https://docs.djangoproject.com/en/3.1/internals/contributing/writing-code/submitting-patches/) every non-trivial changes require [a Trac ticket](https://code.djangoproject.com/).

Beyond the current CI failures and the inconsistencies that this introduces with `add(bulk=False)` which could all be addressed this patch makes an assumption that is backward incompatible and prone to races with regard to local state.

For example, assuming the common `Book -> Author` scenario where `Book` had a foreign key point to `Author`.

If you do something along

```
author = Author.objects.get()
book = Book.objects.get()
author.books.add(book)
```

Then your patch wouldn't perform an update to `book.author` if it's equal to `author`. The thing is nothing guarantees that `book.author` is matching the database state. In other words, `book.author` might have been updated by another process in between the time it was retrieved from the database and `author.add` was called. Applying your proposed changes would make `.add` a noop and impossible to enforce the update of the relationship.

To summarize we can't make assume he local state of ORM objects to match the database and thus cannot proceed with this patch. If you really want to achieve this behaviour, even when aware of its caveats, you can filter objects passed to `add` based on whether or not their relationship is matching.
